### PR TITLE
support merge_type``json``

### LINKED ISSUE #54: Json merge_type is not applying
##### SUMMARY
When using json merge, the playbook is running ok but not doing anything. It is the same as https://github.com/ansible/ansible/issues/65897 that was automatically closed by the repo migration.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
k8s

##### ANSIBLE VERSION

```
ansible 2.9.18
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/rgordill/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.9/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.9.2 (default, Feb 20 2021, 00:00:00) [GCC 10.2.1 20201125 (Red Hat 10.2.1-9)]
```

##### CONFIGURATION

```

```

##### OS / ENVIRONMENT

Fedora release 33 (Thirty Three)
Linux musashi 5.11.11-200.fc33.x86_64 ansible-collections/community.kubernetes#1 SMP Tue Mar 30 16:53:32 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

##### STEPS TO REPRODUCE

1) Start a simple minikube with:
`minikube start -addons ingress`
2) See the configuration of the ingress deployment with
`kubectl get deployments.apps -n kube-system ingress-nginx-controller -o yaml`
3) Execute the following ansible tasks

```yaml
- name: Patch Nginx to support ssl-passthough
  community.kubernetes.k8s:
    state: present
    kind: deployment
    api_version: apps/v1
    name: ingress-nginx-controller
    namespace: kube-system
    merge_type: 
    - json
    definition:
    - op: add
      path: '/spec/template/spec/containers/0/args/-'
      value: '--enable-ssl-passthrough'
```
4) Check with the same `kubectl`command, that the deployment is exactly the same
5) Try to do it manually with
```
kubectl patch deployment \
  ingress-nginx-controller \
  --namespace kube-system \
  --type='json' \
  -p='[{"op": "add", "path": "/spec/template/spec/containers/0/args/-", "value": "--enable-ssl-passthrough"}]'
```
6) Again get the deployment, now the merge has been applied.

##### EXPECTED RESULTS
Merge patch successfully applied as in kubectl.

##### ACTUAL RESULTS

[result.txt](https://github.com/ansible-collections/community.kubernetes/files/6296112/result.txt)

### PR DESCRIPTION
**Pull request by abikouo**
##### SUMMARY

support merge_type ``json``

Fixes #54 
##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

k8s
##### ADDITIONAL INFORMATION

```paste below
- name: Patch Nginx deployment to support ssl-passthough
  kubernetes.core.k8s:
    kind: Deployment
    namespace: "{{ k8s_ns }}"
    name: "{{ k8s_deployment_name }}"
    merge_type: 
    - json
    definition:
    - op: add
      path: '/spec/template/spec/containers/0/args/-'
      value: '--enable-ssl-passthrough'
```

### CODE REVIEW

File: molecule/default/tasks/merge_types.yml [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    kubernetes.core.k8s_info:
+      kind: Deployment
+      name: "{{ k8s_merge_deployment }}"
+      namespace: "{{ k8s_merge_namespace }}"
+    register: describe_depl
+
+  - name: assert that args changed on deployment
+    assert:
+      that:
+      - describe_depl.resources[0].spec.template.spec.containers[0].args | length == 4
```
Comment (gravesm) [inline molecule/default/tasks/merge_types.yml:175]: Could you also add some tests for `merge_type: merge`?

**Review by gravesm — CHANGES_REQUESTED** [review]
Thanks @abikouo! In addition to my comments we need to add jsonpatch to requirements.txt and update the requirements for the k8s module.

File: molecule/default/converge.yml [hunk 2]
```diff
@@ -133,6 +133,14 @@
       tags:
         - always
 
+    - name: Include merge_types.yml
+      include_tasks:
+        file: tasks/merge_types.yml
+        apply:
+          tags: [ merge_type, k8s ]
```
[the remark this answers is not in the document]
Comment (abikouo) [reply molecule/default/converge.yml:140]: agree

File: molecule/default/tasks/merge_types.yml [hunk 3]
```diff
@@ -0,0 +1,175 @@
+- name: Define common facts
```
[the remark this answers is not in the document]
Comment (abikouo) [reply molecule/default/tasks/merge_types.yml]: I was thinking we are running in a crash and burn env, 
but will do it

**Comment by abikouo** [comment]
> Thanks @abikouo! In addition to my comments we need to add jsonpatch to requirements.txt and update the requirements for the k8s module.

I am just wondering as there are some module that are not set as required, as they correspond to a specific functionality
for instance ``yaml`` is used almost everywhere, should this be part of the requirements.txt to ? Just need to clarify the policy

File: molecule/default/tasks/merge_types.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (gravesm) [reply molecule/default/tasks/merge_types.yml]: In CI it's fine, but it depending on how people have things set up locally, it makes it easier.

**Comment by gravesm** [comment]
Yeah, I guess I'm not sure why yaml is not in there. I think jsonpatch should be though, because otherwise the version of k8s that goes into the EE won't be able to use this functionality.

**Review by gravesm — APPROVED** [review]
@Akasurde can you take a look at this?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -87,7 +87,7 @@ jobs:
       # The 3.3.0 release of molecule introduced a breaking change. See
       # https://github.com/ansible-community/molecule/issues/3083
       - name: Install molecule and openshift dependencies
-        run: pip install ansible "molecule<3.3.0" yamllint openshift flake8
+        run: pip install ansible "molecule<3.3.0" yamllint openshift flake8 jsonpatch
```
Comment (goneri) [inline .github/workflows/ci.yml:90]: jsonpatch should also be in `requirements.txt`.

Comment (goneri) [reply .github/workflows/ci.yml:90]: My bad, it's already the case :-).

**Review by goneri — APPROVED** [review]

**Comment by goneri** [comment]
Nice patch, thanks @abikouo.

**Pull request merged**